pipeline/clean/merge-mono.py

```diff
@@ -0,0 +1,234 @@
+import argparse
+import glob
+import json
+import os
+import unicodedata
+from dataclasses import asdict, dataclass
+from pathlib import Path
+from typing import Generator
+
+from pipeline.common.datasets import shuffle_with_max_lines
+from pipeline.common.downloads import read_lines, write_lines
+from pipeline.common.logging import get_logger
+
+logger = get_logger("merge-mono")
+
+# TODO(CJK) - Issue #424
+MAX_WORDS_IN_SENTENCE = 100
+
+
+def hash_line(line: str) -> int:
+    """
+    Return a hash of a line. The line has its whitespace stripped and text representation
+    normalized to ensure a consistent representation.
+    """
+    cleaned_line = unicodedata.normalize("NFC", line.strip())
+    return hash(cleaned_line)
+
+
+@dataclass
+class FilteringStatistics:
+    """
+    Gather statistics about the filtering process.
+    """
+
+    # The size of the merged parallel corpus.
+    parallel_corpus_lines: int = 0
+
+    # How much of the monolingual data was duplicated in the merged parallel corpus.
+    duplicates_of_parallel_corpus: int = 0
+
+    # How much of the monolingual data was duplicated across the monolingual datasets.
+    duplicates_of_monolingual_corpus: int = 0
+
+    # What was the size of the monolingual data that was filtered. This doesn't count the
+    # truncation of datasets at the datasets gathering time.
+    original_monolingual_lines: int = 0
+
+    # After deduplication, how much monolingual data is left.
+    deduplicated_monolingual_lines: int = 0
+
+    # After truncation via the config's `experiment.mono-max-sentences-src.total`,
+    # how many lines are left.
+    final_truncated_monolingual_lines: int = 0
+
+    # The amount of codepoints in the final monolingual corpus.
+    final_truncated_monolingual_codepoints: int = 0
+
+    def save_json(self, path: Path) -> None:
+        with open(path, "w", encoding="utf-8") as json_file:
+            json.dump(asdict(self), json_file, indent=2)
+
+
+def filter_and_write_monolingual_data(
+    mono_datasets: list[str],
+    output_path: Path,
+    parallel_hashes: set[int],
+    max_lines: int,
+    sample_size: int,
+    stats: FilteringStatistics,
+) -> None:
+    """
+    Filtering is done with a set[int]. Seeing if a line is in the set should be O(1)
+    in terms of time complexity. A set[int] was chosen (storing the hash) rather than
+    a set[str], as the latter would retain the string in memory.
+    """
+
+    mono_hashes = set()
+
+    def deduplicate_lines(lines: Generator[str, None, None]) -> Generator[str, None, None]:
```

Comment:
It's getting more complex. Will this code benefit from introducing a class and doing a bit of OOP? You could break the stages to different class methods and store the stats in `self` as well.